File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/comum/validators/UniqueValidator.java
/*
 * Decompiled with CFR 0.152.
 * 
 * Could not load the following classes:
 *  groovy.util.Eval
 */
package br.jus.trt8.pjecalc.negocio.comum.validators;

import br.jus.trt8.pjecalc.base.comum.EntidadeBase;
import br.jus.trt8.pjecalc.base.comum.RepositorioBase;
import br.jus.trt8.pjecalc.base.comum.validadores.CustomValidator;
import br.jus.trt8.pjecalc.base.comum.validadores.ValidatorContext;
import br.jus.trt8.pjecalc.negocio.comum.validators.Unique;
import groovy.util.Eval;

public class UniqueValidator
extends CustomValidator<Unique> {
    @Override
    public Object getParameter(String key) {
        return null;
    }

    @Override
    public String getMessage() {
        return ((Unique)this.getAnnotation()).message();
    }

    @Override
    public byte[] getGroups() {
        return ((Unique)this.getAnnotation()).groups();
    }

    @Override
    protected String getCondition() {
        return ((Unique)this.getAnnotation()).condition();
    }

    private Object getAtributeValue(String name) {
        return Eval.me((String)"bean", (Object)this.getBean(), (String)String.format("bean.%s", name));
    }

    @Override
    public boolean isValid(ValidatorContext context) {
        if (context.getValue() == null) {
            return true;
        }
        Object[] values = new Object[((Unique)this.getAnnotation()).fields().length];
        for (int i = 0; i < values.length; ++i) {
            values[i] = this.getAtributeValue(((Unique)this.getAnnotation()).fields()[i]);
        }
        EntidadeBase entidade = (EntidadeBase)this.getBean();
        return !RepositorioBase.isDuplicado(entidade, ((Unique)this.getAnnotation()).fields(), values);
    }
}

File: src/lib/pjecalc/core/comum/validators/unique-validator.ts
/**
 * PJe-Calc v2.15.1 — UniqueValidator
 * Porte TS-adaptado de: br.jus.trt8.pjecalc.negocio.comum.validators.UniqueValidator
 *
 * No Java consulta `RepositorioBase.isDuplicado` (JPA). Aqui delegamos a
 * função de checagem via callback `duplicateChecker`, que consumidores podem
 * configurar por DI. Sem checker, a validação é sempre ok.
 */
import { CustomValidator } from './custom-validator';
import type { Unique_ } from './unique';
import type { ValidatorContext } from './validator-context';

export type UniqueDuplicateChecker = (
  entidade: unknown,
  fields: string[],
  values: unknown[],
) => boolean;

export class UniqueValidator extends CustomValidator<Unique_> {
  private static duplicateChecker: UniqueDuplicateChecker | null = null;

  static setDuplicateChecker(fn: UniqueDuplicateChecker | null): void {
    UniqueValidator.duplicateChecker = fn;
  }

  getMessage(): string { return this.spec.message; }
  getGroups(): number[] { return this.spec.groups; }
  protected getCondition(): string { return this.spec.condition; }

  private getAttributeValue(name: string): unknown {
    const bean = this.getBean();
    if (bean == null) return null;
    const obj = bean as Record<string, unknown>;
    const getter = `get${name.charAt(0).toUpperCase()}${name.slice(1)}`;
    const fn = obj[getter];
    if (typeof fn === 'function') return (fn as () => unknown).call(obj);
    return obj[name];
  }

  isValid(context: ValidatorContext): boolean {
    if (context.getValue() == null) return true;
    const values = this.spec.fields.map(f => this.getAttributeValue(f));
    if (!UniqueValidator.duplicateChecker) return true;
    return !UniqueValidator.duplicateChecker(this.getBean(), this.spec.fields, values);
  }
}
